legend-engine-language-pure-compiler/src/main/java/org/finos/legend/engine/language/pure/compiler/toPureGraph/Milestoning.java

```diff
@@ -487,4 +495,204 @@ private static Stereotype generatedMilestoningStereotype(CompileContext context,
+    }
+
+    public static boolean isGeneratedMilestoningProperty(AbstractProperty<?> property, CompileContext context)
+    {
+        return isGeneratedMilestoningProperty(property, context, "generatedmilestoningproperty", GENERATED_MILESTONING_PATH_SUFFIX);
+    }
+
+    public static boolean isGeneratedQualifiedProperty(AbstractProperty<?> property, CompileContext context)
+    {
+        return property instanceof QualifiedProperty && isGeneratedMilestoningProperty(property, context) && !isAllVersionsInRangeProperty(property, context);
+    }
+
+    private static int getCountOfParametersSatisfyingMilestoningDateRequirments(QualifiedProperty milestonedQualifiedProperty, CompileContext context)
+    {
+        if (!isGeneratedMilestoningProperty(milestonedQualifiedProperty, context))
+        {
+            throw new EngineException("Unable to get milestoning date parameters for non milestoned QualifiedProperty: " + milestonedQualifiedProperty.getName());
+        }
+        Class returnType = (Class)milestonedQualifiedProperty._genericType()._rawType();
+        MilestoningStereotype milestoningStereotype = Lists.mutable.ofAll(temporalStereotypes(returnType._stereotypes())).getFirst();
+        return 1 + milestoningStereotype.getTemporalDatePropertyNames().size();
+    }
+
+    public static boolean isGeneratedMilestonedQualifiedPropertyWithMissingDates(AbstractProperty<?> property, CompileContext context, Integer parametersCount)
+    {
+        if (isGeneratedQualifiedProperty(property, context))
+        {
+            return parametersCount != getCountOfParametersSatisfyingMilestoningDateRequirments((QualifiedProperty)property, context);
+        }
+        return false;
+    }
+
+    public static boolean isProcessingTemporal(MilestoningStereotype milestoningStereotype)
+    {
+        return milestoningStereotype != null && milestoningStereotype.getPurePlatformStereotypeName() == "processingtemporal";
+    }
+
+    public static boolean isBusinessTemporal(MilestoningStereotype milestoningStereotype)
+    {
+        return milestoningStereotype != null && milestoningStereotype.getPurePlatformStereotypeName() == "businesstemporal";
+    }
+
+    private static boolean isSingleDateTemporal(MilestoningStereotype milestoningStereotype)
+    {
+        return milestoningStereotype != null && isProcessingTemporal(milestoningStereotype) || isBusinessTemporal(milestoningStereotype);
+    }
+
+    private static Type getMilestonedPropertyOwningType(AbstractProperty<?>property)
+    {
+        if (property._owner() instanceof Class)
+        {
+            return (Class) property._owner();
+        }
+        else if (property._owner() instanceof org.finos.legend.pure.m3.coreinstance.meta.pure.metamodel.relationship.Association)
+        {
+            return ((org.finos.legend.pure.m3.coreinstance.meta.pure.metamodel.relationship.Association) property._owner())._originalMilestonedProperties().getFirst()._genericType()._rawType();
```

Comment:
Is this valid if both models of association are milestoned?

Reply:
modified it